File: rust/src/domain/widget_code.rs
use regex::Regex;

#[derive(PartialEq, Debug)]
pub struct WidgetCode(String);

impl WidgetCode {
    pub fn create(code: &str) -> Result<WidgetCode, String> {
        let re = Regex::new(r"W\d{4}").expect(r"W\d{4} is a valid regex");
        if re.is_match(code) {
            Ok(WidgetCode(String::from(code)))
        } else {
            Err(String::from(
                "`WidgetCode` must begin with a 'W' and be followed by 4 digits",
            ))
        }
    }

    pub fn value(&self) -> &str {
        &self.0
    }
}

File: fsharp/dmmf/ordering/WidgetCode.fs
namespace Ordering

open System.Text.RegularExpressions

type WidgetCode = private WidgetCode of string
module WidgetCode =
    let create (code : string) : Result<WidgetCode, string> =
        if Regex.IsMatch(code, @"W\d{4}") then
            Ok (WidgetCode code)
        else
            Error "`WidgetCode` must begin with a 'W' and be followed by 4 digits"

    let value (WidgetCode code) = code
